File: Java/Reverse Words in a String II.java
/*
Similar to Question [6. Reverse Words in a String], but with the following constraints:
“The input string does not contain leading or trailing spaces and the words are always separated by a single space.”
Could you do it in-place without allocating extra space?
*/

//O(n) runtime, O(1) space – In-place reverse:
public void reverseWords(char[] s) {
   reverse(s, 0, s.length);
   for (int i = 0, j = 0; j <= s.length; j++) 
   {
      if (j == s.length || s[j] == ' ') {
         reverse(s, i, j);
         i = j + 1;
	} 
}
}

private void reverse(char[] s, int begin, int end) {
   for (int i = 0; i < (end - begin) / 2; i++) {
      char temp = s[begin + i];
      s[begin + i] = s[end - i - 1];
      s[end - i - 1] = temp;
} 
}

File: Python/Reverse Words in a String II.py
"""
Given an input string, reverse the string word by word. A word is defined as a sequence of non-space characters.

The input string does not contain leading or trailing spaces and the words are always separated by a single space.

For example,
Given s = "the sky is blue",
return "blue is sky the".

Could you do it in-place without allocating extra space?

Related problem: Rotate Array
"""

class Solution(object):
    def reverseWords(self, s):
        """
        :type s: a list of 1 length strings (List[str])
        :rtype: nothing
        """
        self.reverse(s, 0, len(s) - 1)
        start = 0 
        for i in xrange(len(s)):
            if s[i] == ' ':
                self.reverse(s, start, i - 1)
                start = i + 1
            elif i == len(s) - 1:
                self.reverse(s, start, i)
                
    def reverse(self, s, start, end):
        while start < end:
            s[start], s[end] = s[end], s[start]
            start += 1
            end -= 1
